src/Conversion/ONNXToKrnl/Tensor/Unique.cpp

```diff
@@ -0,0 +1,194 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===------------------- Unique.cpp - Lowering Unique Op ----------------===//
+//
+// Copyright 2019-2022 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers the ONNX Unique Operator to Krnl dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "onnx-mlir/Runtime/OMTensor.h"
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+#include "src/Dialect/ONNX/ONNXOps/ShapeHelper.hpp"
+#include "llvm/Support/Debug.h"
+using namespace mlir;
+
+namespace onnx_mlir {
+
+/// Emit function call to compute arg unique of a given MemRef along a given
+/// axis. The first output MemRef has the same shape as the input MemRef but is
+/// of IndexType. Shape of the second, third and fourth arguments depends on the
+/// input options.
+Value emitArgUnique(ConversionPatternRewriter &rewriter, Location loc,
+    Value total, Value input, int64_t axis, int64_t sorted, Value Y,
+    Value indices, Value inverse_indices, Value counts, bool count_only) {
+  MultiDialectBuilder<KrnlBuilder, IndexExprBuilderForKrnl, MathBuilder> create(
+      rewriter, loc);
+  IndexExprScope scope(create.krnl);
+  MemRefType inputMemRefType = input.getType().cast<MemRefType>();
+  int64_t rank = inputMemRefType.getRank();
+  assert(axis < rank && "axis is out of bound");
+  LiteralIndexExpr zeroIE(0), oneIE(1);
+  SmallVector<IndexExpr, 4> lbs(rank, zeroIE);
+  SmallVector<IndexExpr, 4> ubs;
+  create.krnlIE.getShapeAsDims(input, ubs);
+
+  // Emit krnl.Call to call omTensorUnique API
+  Type int_type = rewriter.getIntegerType(64);
+  Value val_axis = create.math.constant(int_type, axis);
+  Value val_sorted = create.math.constant(int_type, sorted);
+  if (count_only) {
+    SmallVector<Value, 4> operands = {total, input, val_axis, val_sorted};
+    rewriter.create<KrnlCallOp>(loc, "omTensorUniqueCount", 1, operands);
+  } else {
+    SmallVector<Value, 8> operands = {total, input, val_axis, val_sorted, Y,
+        indices, inverse_indices, counts};
+    rewriter.create<KrnlCallOp>(loc, "omTensorUnique", 8, operands);
```

Comment:
Is `8` the correct number for outputs?

Reply:
I changed the number from 8 to 5.  I changed the argument order of omTensorUnique to gather output arguments at the beginning. Thanks.